modules/utxo-lib/src/bitgo/Musig2.ts

```diff
@@ -0,0 +1,154 @@
+  pubNonces: Buffer
+): PsbtProprietaryKeyValueData {
+  if (tapOutputKey.length !== 32) {
+    throw new Error(`Invalid tapOutputKey length ${tapOutputKey.length}`);
+  }
+  if (pubNonces.length !== 66) {
+    throw new Error(`Invalid pubNonces length ${pubNonces.length}`);
+  }
+  const keydata = Buffer.alloc(96);
+  toXOnlyPublicKey(participantKey).copy(keydata);
+  tapOutputKey.copy(keydata, 32);
+  const key = {
+    identifier: PSBT_IN_TAP_MUSIG2_PUB_NONCE,
+    subtype: PSBT_IN_TAP_MUSIG2_PUB_NONCE_SUBTYPE,
+    keydata,
+  };
+  return { key, value: pubNonces };
+}
+
+export function generateMusig2Nonce(psbt: UtxoPsbt, rootWalletKey: BIP32Interface, sessionId?: Buffer): void {
+  if (sessionId && sessionId.length > Math.pow(2, 32) - 1) {
+    throw new Error('Invalid sessionId size. Max supported size is 2^32 - 1');
+  }
+
+  psbt.data.inputs.forEach((input, inputIndex) => {
+    if (!input.tapInternalKey) {
+      // not a p2tr musig2 key path spending.
+      return;
+    }
+
+    if (!input.tapBip32Derivation || input.tapBip32Derivation.length === 0) {
+      throw new Error('Need tapBip32Derivation to generate musig2 nonce');
+    }
+
+    const tapInternalKeyIndex = input.tapBip32Derivation.findIndex((bipDv) => {
+      return input.tapInternalKey!.equals(bipDv.pubkey);
+    });
+    if (tapInternalKeyIndex >= 0) {
+      // two possible cases after this:
+      // 1. rootWalletKey derives tapInternalKey - it means non-aggregated tapInternalKey
+      // 2. rootWalletKey doesn't derive tapInternalKey
+      // Neither concerns this function, so skip it.
+      return;
+    }
+
+    const participantsPubkeyKeyValues = psbt.getProprietaryKeyVals(inputIndex, {
+      identifier: PSBT_IN_MUSIG_PARTICIPANT_PUB_KEYS,
+      subtype: PSBT_IN_MUSIG_PARTICIPANT_PUB_KEYS_SUBTYPE,
+    });
+    if (participantsPubkeyKeyValues.length !== 1) {
+      throw new Error(
+        `Invalid number of participant pub keys keydata ${participantsPubkeyKeyValues.length} found instead of 1`
+      );
+    }
+
+    const participantsPubkeyKeyValue = participantsPubkeyKeyValues[0];
+    const participantsPubkeyKeydata = participantsPubkeyKeyValue.key.keydata;
+    if (participantsPubkeyKeydata.length !== 64) {
+      throw new Error(`Invalid keydata size ${participantsPubkeyKeydata.length} for participant pub keys`);
+    }
+    if (!participantsPubkeyKeydata.includes(input.tapInternalKey, 32)) {
+      throw new Error(`No tapInternalKey match with participant pub keys keydata`);
+    }
+    if (participantsPubkeyKeyValue.value.length !== 64) {
+      throw new Error(`Invalid valuedata size ${participantsPubkeyKeyValue.value.length} for participant pub keys`);
+    }
+    const tapOutputKey = participantsPubkeyKeydata.subarray(32);
+
+    const participantPubKeys = [0, 1].map((i) => {
+      return participantsPubkeyKeyValue.value.subarray(i * 32, i * 32 + 32);
+    });
+    if (new Set(participantPubKeys).size !== participantPubKeys.length) {
```

Comment:
duplicate keys are allowed by the musig spec, but I'm fine with us not accepting them.